README.md
= flog

home :: http://ruby.sadi.st/
code :: https://github.com/seattlerb/flog
rdoc :: http://seattlerb.rubyforge.org/flog
vim  :: http://github.com/sentientmonkey/vim-flog

== DESCRIPTION:

Flog reports the most tortured code in an easy to read pain
report. The higher the score, the more pain the code is in.

== FEATURES/PROBLEMS:

* Easy to read reporting of complexity/pain.

== SYNOPSIS:

    % ./bin/flog -g lib
    Total Flog = 1097.2 (17.4 flog / method)
    
       323.8: Flog total
        85.3: Flog#output_details
        61.9: Flog#process_iter
        53.7: Flog#parse_options
    ...

== REQUIREMENTS:

* ruby2ruby (only for -v)
* ruby_parser

== INSTALL:

* sudo gem install flog

== LICENSE:

(The MIT License)

Copyright (c) Ryan Davis, seattle.rb

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
=== 4.3.0 / 2014-07-18

* 1 minor enhancement:

  * Added methods_only param to FlogTask. (jocranford)

=== 4.2.1 / 2014-05-29

* 2 bug fixes:

  * Fixed grouping by class name. (guilhermesimoes)
  * Removed dead rubyforge setting in Rakefile

=== 4.2.0 / 2013-10-18

* 3 minor enhancements:

  * Added --extended / -e to put file:line output on next line (for rubymine & friends)
  * Extracted MethodBasedSexpProcessor and pushed up to sexp_processor gem.
  * Updated dependency on sexp_processor to ~> 4.4.

=== 4.1.2 / 2013-09-05

* 2 bug fixes:

  * Don't even record a score if method stack is empty and user used --methods-only.
  * Fixed Flog task's default method. (jasonrobertfox)

=== 4.1.1 / 2013-07-11

* 1 minor enhancement:

  * Added max_method to delegators.

* 1 bug fix:

  * Fixed FlogTask to use FlogCLI. (envygeeks)

=== 4.1.0 / 2013-05-10

* 3 minor enhancements:

  * Cleaned up tests by adding assert_hash_in_epsilon. yay!
  * Fixed method_location is now cleared on #reset. (makaroni4)
  * to_proc_normal is now penalized based on RUBY_VERSION. Scores were benchmarked.

* 1 bug fix:

  * Fixed code/home urls in readme/gem.

=== 4.0.0 / 2013-04-18

* 1 major enhancement:

  * Renamed Flog#methods to #method_scores. (makaroni4)

* 4 minor enhancements:

  * Added accessors for methods, scores. Now available for CIs! (makaroni4)
  * Refactored calculations down to 2 methods: total_score and totals. (makaroni4)
  * Refactored to #calculate, #threshold. (makaroni4)
  * Track stack of nested sclass scopes for parser (pithyless)

=== 3.2.3 / 2013-03-21

* 1 bug fix:

  * Don't blow up if a path doesn't exist or isn't readable.

=== 3.2.2 / 2013-01-22

* 1 bug fix:

  * Fixed stupid error when run on empty file.

=== 3.2.1 / 2012-12-19

* 1 bug fix:

  * Relaxed (and fixed) the ruby_parser dependency.

=== 3.2.0 / 2012-12-18

* 4 minor enhancements:

  * Ensure rake/tasklib is loaded when defining FlogTask. (ferrous26)
  * Fixed reporting / recording of methods in singleton class blocks. (mihu)
  * Refactored error handling code to flog_ruby. Calls flog_ruby! for actual work
  * Refactored ruby processing code to flog_ruby!

* 1 bug fix:

  * Fixed flogging of || iters. (JacobNinja)

=== 3.1.0 / 2012-11-16

* 4 minor enhancements:

  * --quiet option is now false by default, and changed to silence parse warnings.
  * Added max_score and max_method. (aselder)
  * FlogTask can now take a method to use to figure out score, allowing for total or max_score.
  * Switched to capturing RubyParser::SyntaxError (RP 3.0 change).

* 2 bug fixes:

  * Avoid redefined warning for File::RUBY19. (svendahlstrand)
  * Fixed flog to default to RubyParser if not specified. Allows cleaner programmatic access.

=== 3.0.0 / 2012-11-02

* 1 minor enhancement:

  * Added a timeout handler to skip when RubyParser times out on a large file

* 1 bug fix:

  * Fixed handling of plain literals in masgn in args.

=== 3.0.0.b3 / 2012-10-22

* 4 minor enhancements:

  * Added .rake as a supported extension.
  * Create a new parser for every file, preventing state barkification
  * Extended DSL support to include hash args (eg task :blah => :dep).
  * Extended DSL support to report nested names (eg namespace(blah)::task#woot)

=== 3.0.0.b2 / 2012-08-07

* 1 bug fix:

  * Fix for sexp structure changes (no arglists).

=== 3.0.0.b1 / 2012-07-26

* 4 minor enhancements:

  * Added --18 and --19 flags to specify parser. Defaults to hybrid.
  * Explicitly use Ruby18Parser to remove deprecation warnings.
  * Modified processor to deal with cleaner sexps from RP 3.x.
  * Use File.binread (File.read in 1.8) to bypass encoding errors

* 2 bug fixes:

  * Cleaned up some 1.9 warnings.
  * Fixed failing tests against ruby_parser 3

=== 2.5.3 / 2011-09-21

* 1 minor enhancement:

  * class_stack now always pushes on symbols or strings for class names.

* 1 bug fix:

  * Fixed klass_name returning an ugly mix of sexps and strings.

=== 2.5.2 / 2011-08-11

* 1 minor enhancement:

  * Improved parse error output. (impurist)

=== 2.5.1 / 2011-02-18

* 3 minor enhancements:

  * Added RegexpError to error handler.
  * Improved error output and suggest --continue
  * Record the flog score for the full class name, not just suffix. (dkubb)

* 1 bug fix:

  * Fixed block_pass when passed a multi-level const (xavier)

=== 2.5.0 / 2010-09-01

* 1 major enhancement:

  * Added plugin system. Define a module under Flog to extend it.

* 3 minor enhancements:

  * Added special case penalty for wtf to_proc: blah(&b = proc {...}) (benjaminb)
  * Improved tests and test coverage.
  * Unfactored & refactored report code. Much cleaner and  more maintainable now.

* 2 bug fixes:

  * Fixed API change for FlogTask (andreacampi)
  * Fixed bad edgecase handler for block_pass (benjaminb)

=== 2.4.0 / 2009-12-15

* 4 minor enhancements:

  * Cleaned method_name to return "#method" or "::method".
  * DSL reporting now handles regexp literals for 'method' names (Marty Andrews)
  * Improved tests for process_iter's myriad complexities.
  * More doco!

* 1 bug fix:

  * Fixed DSL reporting excluding solo-blocks. (eg a single rake task)

=== 2.3.0 / 2009-12-09

* 1 major enhancement:

  * Added file:line info to the flog report (Marty Andrews)

* 13 minor enhancements:

  * Added .autotest.
  * Deleted pre-gauntlet scripts.
  * Flog#method_name now at least tries show when it is a class method.
  * Flog.parse_options now takes args directly.
  * Removed Flog#increment_total_score_by.
  * Removed Flog#output_summary.
  * Removed Flog#process_attrset.
  * Removed Flog#record_method_score.
  * Removed Flog#summarize_method.
  * Removed Flog::default_options.
  * Renamed Flog#analyze_list to process_until_empty.
  * Renamed Flog#options to #option
  * Rewrote entire test suite (3100 lines down!). Cleaner and less brittle.

=== 2.2.0 / 2009-08-14

* 1 minor enhancement:

  * #mass pushed up to sexp_processor

* 5 bug fixes:


lib/flog_cli.rb
require "rubygems"
require "optparse"
require "forwardable"

require "path_expander"
require "flog"

class FlogCLI
  extend Forwardable

  def_delegators :@flog, :average, :calculate, :each_by_score, :option
  def_delegators :@flog, :method_locations, :method_scores, :reset, :scores
  def_delegators :@flog, :threshold, :total_score, :no_method, :calculate_total_scores
  def_delegators :@flog, :max_method

  def self.run args = ARGV
    load_plugins

    expander = PathExpander.new args, "**/*.{rb,rake}"
    files = expander.process

    options = parse_options args

    flogger = new options
    flogger.flog(*files)
    flogger.report
  end

  ##
  # Loads all flog plugins. Files must be named "flog/*.rb".

  def self.load_plugins
    # TODO: I think I want to do this more like hoe's plugin system. Generalize?
    loaded, found = {}, {}

    Gem.find_files("flog/*.rb").reverse.each do |path|
      found[File.basename(path, ".rb").intern] = path
    end

    found.each do |name, plugin|
      next if loaded[name]
      begin
        warn "loading #{plugin}" # if $DEBUG
        loaded[name] = load plugin
      rescue LoadError => e
        warn "error loading #{plugin.inspect}: #{e.message}. skipping..."
      end
    end

    self.plugins.merge loaded

    names = Flog.constants.map {|s| s.to_s}.reject {|n| n =~ /^[A-Z_]+$/}

    names.each do |name|
      # next unless Hoe.plugins.include? name.downcase.intern
      mod = Flog.const_get(name)
      next if Class === mod
      warn "extend #{mod}" if $DEBUG
      # self.extend mod
    end
  end

  ##
  # Parse options in +args+ (defaults to ARGV).

  def self.parse_options args = ARGV
    option = {
      :quiet    => false,
      :continue => false,
      :parser   => RubyParser,
    }

    OptionParser.new do |opts|
      opts.separator "Standard options:"

      opts.on("-a", "--all", "Display all flog results, not top 60%.") do
        option[:all] = true
      end

      opts.on("-b", "--blame", "Include blame information for methods.") do
        option[:blame] = true
      end

      opts.on("-c", "--continue", "Continue despite syntax errors.") do
        option[:continue] = true
      end

      opts.on("-d", "--details", "Show method details.") do
        option[:details] = true
      end

      opts.on("-g", "--group", "Group and sort by class.") do
        option[:group] = true
      end

      opts.on("-h", "--help", "Show this message.") do
        puts opts
        exit
      end

      opts.on("-I dir1,dir2,dir3", Array, "Add to LOAD_PATH.") do |dirs|
        dirs.each do |dir|
          $: << dir
        end
      end

      opts.on("-m", "--methods-only", "Skip code outside of methods.") do
        option[:methods] = true
      end

      opts.on("-q", "--quiet", "Don't show parse errors.") do
        option[:quiet] = true
      end

      opts.on("-e", "--extended", "Put file:line on a separate line (for rubymine & friends).") do
        option[:extended] = true
      end

      opts.on("-s", "--score", "Display total score only.") do
        option[:score] = true
      end

      opts.on("-tN", "--threshold=N", Integer, "Set the report cutoff threshold (def: 60%).") do |n|
        option[:threshold] = n / 100.0
      end

      opts.on("-v", "--verbose", "Display progress during processing.") do
        option[:verbose] = true
      end

      opts.on("--18", "Use a ruby 1.8 parser.") do
        option[:parser] = Ruby18Parser
      end

      opts.on("--19", "Use a ruby 1.9 parser.") do
        option[:parser] = Ruby19Parser
      end

      next if self.plugins.empty?
      opts.separator "Plugin options:"

      extra = self.method_scores.grep(/parse_options/) - %w(parse_options)

      extra.sort.each do |msg|
        self.send msg, opts, option
      end

    end.parse! Array(args)

    option
  end

  ##
  # The known plugins for Flog. See Flog.load_plugins.

  def self.plugins
    @plugins ||= {}
  end

  ##
  # Flog the given files. Deals with "-", syntax errors, and
  # traversing subdirectories intelligently. Use PathExpander to
  # process dirs into files.

  def flog(*files)
    files << "-" if files.empty?
    @flog.flog(*files)
  end

  ##
  # Creates a new Flog instance with +options+.

  def initialize options = {}
    @flog = Flog.new options
  end

  ##
  # Output the report up to a given max or report everything, if nil.

  def output_details io, max = nil
    io.puts

    each_by_score max do |class_method, score, call_list|
      self.print_score io, class_method, score

      if option[:details] then
        call_list.sort_by { |k,v| -v }.each do |call, count|
          io.puts "  %6.1f:   %s" % [count, call]
        end
        io.puts
      end
    end
  end

  ##
  # Output the report, grouped by class/module, up to a given max or
  # report everything, if nil.

  def output_details_grouped io, threshold = nil
    calculate

    scores.sort_by { |_, n| -n }.each do |klass, total|
      io.puts

      io.puts "%8.1f: %s" % [total, "#{klass} total"]

      method_scores[klass].each do |name, score|
        self.print_score io, name, score
      end
    end
  end

  ##
  # Print out one formatted score.

  def print_score io, name, score
    location = method_locations[name]
    if location then
      sep = " "
      sep = "%-11s" % "\n" if option[:extended]
      io.puts "%8.1f: %-32s%s%s" % [score, name, sep, location]
    else
      io.puts "%8.1f: %s" % [score, name]
    end
  end

  ##
  # Report results to #io, STDOUT by default.

  def report(io = $stdout)
    io.puts "%8.1f: %s" % [total_score, "flog total"]
    io.puts "%8.1f: %s" % [average, "flog/method average"]

    return if option[:score]

    if option[:group] then
      output_details_grouped io, threshold
    else
      output_details io, threshold
    end
  ensure
    self.reset
  end
end
